File: diplomacy2/submissions/partially_accepted/line_st.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define com complex<double>
#define ld long double;
#define ii pair<ll,ll>
#define vi vector<int>
#define vii vector<ii>
#define vc vector<char>
#define vs vector<string>
#define vd vector<double>
#define vcom vector<com>
#define vld vector<ld>
#define vll vector<ll>
#define vvi vector<vi>
#define vvii vector<vii>
#define vvc vector<vc>
#define vvs vector<vs>
#define vvll vector<vll>
#define vvd vector<vd>
#define vvcom vector<vcom>
#define vvld vector<vld>
#define FOR(x,n) for(int x=0;x<(n);x++)
#define FORS(x,n) for(int x=1;x<=(n);x++)
#define FORE(x,a) for(auto &x: a)
#define FORT(x,a,b) for(int x=(a);x<(b);x++)
#define FORD(x,a,b) for(int x=(a);x>=(b);x--)
#define ALL(x) x.begin(),x.end()
#define REP(n) for(int _ = 0; _ < n; _++)
#define MT make_tuple
#define MP make_pair
#define pb push_back
#define endl '\n'
#define F first
#define S second
#define vvvll vector<vvll>
#define sz(x) ((int)x.size())

const int inf = 1e9;
struct FT {
    int n;
	vector<ll> s;
	FT(int _n) : n(_n), s(n+1) {}
	void update(int x,int val) { 
        x++;  
        while(x<=n) {
            s[x]+=val;
            x+=(x&-x);  
        } 
    }

	ll query(int x) {
		ll res = 0;
        x++;
        while(x > 0) {
            res += s[x];
            x -= x&-x;
        }
		return res;
	}
};

struct ST {
    int n;
    int tz;
    vi t, lazy;
    ST(int _n) : n(_n), tz(4*n), t(tz), lazy(tz,1) {}
    
    void push(int v) {
        t[v*2] *= lazy[v];
        lazy[v*2] *= lazy[v];
        t[v*2+1] *= lazy[v];
        lazy[v*2+1] *= lazy[v];
        lazy[v] = 1;
    }
    
    void update(int v, int tl, int tr, int i, int inc) {
        if (i < tl || i > tr) return;
        if (tl == i && tr == i) {
            t[v] += inc;
        } else {
            push(v);
            int tm = (tl + tr) / 2;
            update(v*2, tl, tm, i, inc);
            update(v*2+1, tm+1, tr, i, inc);
            t[v] = t[v*2] + t[v*2+1];
        }
    }
    void update(int i, int inc) {
        update(1,0,n-1,i,inc);
    }
    
    void flip(int v, int tl, int tr, int l, int r) {
        if (tl > r || tr < l) return;
        if (l <= tl && tr <= r) {
            lazy[v] *= -1;
            t[v] *= -1;
        } else {
            push(v);
            int tm = (tl + tr) / 2;
            flip(v*2, tl, tm, l, r);
            flip(v*2+1, tm+1, tr, l, r);
            t[v] = t[v*2] + t[v*2+1];
        }
    }
    void flip(int l, int r) {
        flip(1,0,n-1,l,r);
    }
    
    int query() {
        return t[1];
    }
};


FT cnt(0);
ST D(0);
int n,q;
void mal(int sz,int cf) {
    int ttl = cnt.query(n);
    int sml = cnt.query(sz);
    int cc = (1-2*((ttl-sml)%2))*sz;
    cnt.update(sz,cf);

    D.flip(0,sz);
    D.update(sz,cc);
}

int main() {
    cin >> n >> q;
    cnt = FT(n+1);
    D = ST(n+1);
    cnt.update(1,n);
    map<int,int> st;
    priority_queue<tuple<int,int,int>> pq;
    FOR(i,n) st[i] = i;

    if(n % 2) {
        D.update(1,1);
    }
    FOR(i,q) {
        string c;
        cin >> c;
        if(c == "a") {
            int u,v,p;
            cin >> u >> v >> p;
            u--;
            v--;
            if(u + 1 != v) return 0;
            int pu = st.lower_bound(u)->first;
            int pv = st.lower_bound(v)->first;
            if(pu != pv) {
                int su = pu-st[pu] + 1;
                int sv = pv-st[pv] + 1;
                st[pv] = st[pu];
                st.erase(pu);
                mal(su,-1);
                mal(sv,-1);
                mal(su+sv,1);
            } 
            pq.push({p,u,v});
        } else if(c=="r") {
            int u,v,p;
            tie(p,u,v) = pq.top();
            pq.pop();
            int pp = st.lower_bound(v)->first;
            st[u] = st[pp];
            st[pp] = v;
            int su = pp-st[pp] + 1;
            int sv = u-st[u] + 1;
            int ts = su+sv;
            mal(ts,-1);
            mal(su,1);
            mal(sv,1);
        } else {
            cout << D.query() << endl;
        }
    }
}

File: diplomacy2/submissions/partially_accepted/line_st.py
import heapq
class FenwickTree:
    def __init__(self, size):
        self.n = size
        self.s = [0] * (self.n + 2)  # 1-based indexing

    def update(self, x, val):
        x += 1  # Convert to 1-based index
        while x <= self.n:
            self.s[x] += val
            x += x & -x

    def query(self, x):
        x += 1
        res = 0
        while x > 0:
            res += self.s[x]
            x -= x & -x
        return res

    def lower_bound(self, sum):
        if sum <= 0: 
            return -1
        pos = 0
        pw = 1 << 25
        while pw:
            if pos + pw <= self.n and self.s[pos + pw] < sum:
                pos += pw
                sum -= self.s[pos]
            pw //= 2
        return pos

class SegmentTree:
    def __init__(self, size):
        self.n = size
        self.t = [0] * (4 * self.n)
        self.lazy = [1] * (4 * self.n)

    def push(self, v):
        if self.lazy[v] != 1:
            self.t[2*v] *= self.lazy[v]
            self.lazy[2*v] *= self.lazy[v]
            self.t[2*v+1] *= self.lazy[v]
            self.lazy[2*v+1] *= self.lazy[v]
            self.lazy[v] = 1

    def update_point(self, v, tl, tr, pos, inc):
        if tl == tr:
            self.t[v] += inc
        else:
            self.push(v)
            tm = (tl + tr) // 2
            if pos <= tm:
                self.update_point(2*v, tl, tm, pos, inc)
            else:
                self.update_point(2*v+1, tm+1, tr, pos, inc)
            self.t[v] = self.t[2*v] + self.t[2*v+1]

    def flip_range(self, v, tl, tr, l, r):
        if tr < l or tl > r:
            return
        if l <= tl and tr <= r:
            self.t[v] *= -1
            self.lazy[v] *= -1
        else:
            self.push(v)
            tm = (tl + tr) // 2
            self.flip_range(2*v, tl, tm, l, r)
            self.flip_range(2*v+1, tm+1, tr, l, r)
            self.t[v] = self.t[2*v] + self.t[2*v+1]

    def update(self, pos, inc):
        self.update_point(1, 0, self.n-1, pos, inc)

    def flip(self, l, r):
        self.flip_range(1, 0, self.n-1, l, r)

    def query_total(self):
        return self.t[1]




n,q = map(int,input().split())

cnt = FenwickTree(n+1)
cnt.update(1, n)  # Initially n intervals of size 1

D = SegmentTree(n+1)
if n % 2 == 1:
    D.update(1, 1)

pq = []
it = FenwickTree(n+1)
st = list(range(n))

for i in range(n):
    it.update(i,1)

def mal(sz,cf):
    ttl = cnt.query(n)
    sml = cnt.query(sz)
    cc = (1 - 2 * ((ttl - sml) % 2)) * sz
    cnt.update(sz, cf)
    D.flip(0, sz)
    D.update(sz, cc)

ccnt = n

for _ in range(q):
    s = input().split()
    cmd = s[0]
    if cmd == 'a':
        u,v,p = map(int,s[1:])
        u -= 1
        v -= 1
        if u +1 != v:
            continue
        pu = it.lower_bound(it.query(u-1) + 1)
        pv = it.lower_bound(it.query(v-1) + 1)

        if pu != pv:
            ccnt -= 1
            su = pu - st[pu] + 1
            sv = pv - st[pv] + 1
            it.update(pu,-1)
            
            st[pv] = st[pu]
            st[pu] = -1

            mal(su, -1)
            mal(sv, -1)
            mal(su+sv, 1)
        heapq.heappush(pq, (-p, u, v))
    elif cmd == 'r':
        ccnt += 1
        p_neg, u, v = heapq.heappop(pq)
        p = -p_neg

        pp = it.lower_bound(it.query(u) + 1)

        it.update(u,1)
        st[u] = st[pp]
        st[pp] = v

        su = u-st[u] + 1
        sv = pp-st[pp] + 1
        ts = su + sv
        mal(ts,-1)
        mal(su,1)
        mal(sv,1)
    elif cmd == 'd':
        print(D.query_total(),flush=False)
